core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -32,208 +38,256 @@
+          this(memberId, startOffset, lastOffset, state, 0);
       }
-    }
 
-    public int deliveryCount() {
-      return deliveryCount;
-    }
+      public InFlightFetchBatch(String memberId, long startOffset, long lastOffset, RecordState state, int deliveryCount) {
+          this.memberId = memberId;
+          this.startOffset = startOffset;
+          this.lastOffset = lastOffset;
+          this.state = state;
+          this.deliveryCount = deliveryCount;
+      }
 
-    @Override
-    public int hashCode() {
-      return Objects.hash(state, deliveryCount);
-    }
+      public String memberId() {
+          return memberId;
+      }
 
-    @Override
-    public boolean equals(Object obj) {
-      if (this == obj) {
-        return true;
+      public long startOffset() {
+          return startOffset;
       }
-      if (obj == null) {
-        return false;
+
+      public long lastOffset() {
+          return lastOffset;
       }
-      if (getClass() != obj.getClass()) {
-        return false;
+
+      public RecordState messageState() {
+          return state;
       }
-      InFlightRecord other = (InFlightRecord) obj;
-      return state == other.state && deliveryCount == other.deliveryCount;
-    }
 
-    @Override
-    public String toString() {
-      return "InFlightRecord(" +
-          " state=" + state.toString() +
-          ", deliveryCount=" + ((deliveryCount == 0) ? "" : ("(" + deliveryCount + ")")) +
-          ")";
-    }
+      public int deliveryCount() {
+          return deliveryCount;
+      }
+
+      public List<Long> gapOffsets() {
+          return gapOffsets;
+      }
+
+      public void setMessageState(RecordState newState, boolean incrementDeliveryCount) {
+          state = newState;
+          if (incrementDeliveryCount) {
+              deliveryCount++;
+          }
+      }
+
+      public void addGapOffsets(List<Long> gapOffsets) {
+          if (this.gapOffsets == null) {
+              this.gapOffsets = new ArrayList<>(gapOffsets);
+          } else {
+              this.gapOffsets.addAll(gapOffsets);
+              // Maintain the order of the gap offsets as out of band acknowledgements can be received.
+              Collections.sort(this.gapOffsets);
+          }
+      }
+
+      @Override
+      public String toString() {
+          return "InFlightFetchBatch(" +
+              " memberId=" + memberId +
+              ", startOffset=" + startOffset +
+              ", lastOffset=" + lastOffset +
+              " state=" + state.toString() +
+              ", deliveryCount=" + ((deliveryCount == 0) ? "" : ("(" + deliveryCount + ")")) +
+              ", gapOffsets=" + ((gapOffsets == null) ? "" : gapOffsets) +
+              ")";
+      }
   }
 
   static class AcknowledgementBatch {
 
-    private final long startOffset;
-    private final long lastOffset;
-    private final List<Long> gapOffsets;
-    private final byte acknowledgeType;
+      private final long startOffset;
```

Comment:
This is a mistake in the request schema. Please use `baseOffset` for this variable. I'll update the schema next time we do a refresh.

Reply:
Done.